src/NuGet.Core/NuGet.Packaging/Signing/TrustStore/CertificateBundleX509ChainFactory.cs

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#if NET5_0_OR_GREATER
+
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+
+namespace NuGet.Packaging.Signing
+{
+    internal abstract class CertificateBundleX509ChainFactory : IX509ChainFactory
+    {
+        public X509Certificate2Collection Certificates { get; }
+        public string FilePath { get; }
+
+        protected CertificateBundleX509ChainFactory(X509Certificate2Collection certificates, string filePath = null)
+        {
+            Certificates = certificates;
+            FilePath = filePath;
+        }
+
+        public X509Chain Create()
+        {
+            X509Chain x509Chain = new();
+
+            x509Chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
+
+            if (Certificates is not null && Certificates.Count > 0)
+            {
+                x509Chain.ChainPolicy.CustomTrustStore.AddRange(Certificates);
+            }
+
+            return x509Chain;
+        }
+
+        protected static bool TryImportFromPemFile(string filePath, out X509Certificate2Collection certificates)
+        {
+            certificates = new X509Certificate2Collection();
+
+            try
+            {
+                certificates.ImportFromPemFile(filePath);
+
+                return true;
+            }
+            catch (Exception ex) when (ex is CryptographicException || ex is IOException)
```

Comment:
Just curious why you're swallowing specific exception types and allowing others to throw?  `ImportFromPemFile` claims it will only throw `CryptographicException` and `ArgumentNullException`.

https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2collection.importfrompemfile?view=net-6.0#exceptions

If `filePath` happens to be `null`, an `ArgumentNullException` will be thrown and `certificates.Clear()` won't run but maybe we want that?

Reply:
If you look at the **Remarks** section in the link that you provided, you'll see:

> See [ReadAllText(String)](https://docs.microsoft.com/en-us/dotnet/api/system.io.file.readalltext?view=net-6.0#system-io-file-readalltext(system-string)) for additional documentation about exceptions that can be thrown.

So, for example, if the file does not exist at the provided file path, `ImportFromPemFile(...)` will throw a `FileNotFoundException`, which derives from `IOException`.

There's no scenario in production code where `filePath` will be `null` here.

Reply:
Okay so there's more potential exceptions that could be thrown, can you expand on why only `CryptographicException` and `IOException` are worth catching?  ie `UnauthorizedAccessException`, `SecurityException`

I'd also like to understand why IO exceptions are just swallowed?  If a path is too long for example, the code will just fail silently.  I guess I just want to understand the intent here. the method appears to just wrap `ImportFromPemFile` and return true or false.  Does the caller log something useful if it fails?

Reply:
I narrowed `IOException` down to `DirectoryNotFoundException` and `FileNotFoundException`.

On Linux, `TryImportFromPemFile(...)` will try to load the following files:

*  /etc/pki/ca-trust/extracted/pem/objsign-ca-bundle.pem.
*  codesignctl.pem.  This file will always be located in the versioned SDK directory alongside this assembly (NuGet.Packaging.dll).

Given this:

1.  `PathTooLongException` is unlikely.
2.  `UnauthorizedAccessException` and `SecurityException` are unlikely, because both locations should be readable by a user.
3.  `DirectoryNotFoundException` and `FileNotFoundException` are both likely.

`CryptographicException` covers an assortment of cases, such as unsupported cryptographic algorithms, decoding errors, tampered/invalid signatures, and more.

The caller (`X509TrustStore`) logs which trust store was selected for use.